criu/files-reg.c

```diff
@@ -321,19 +321,48 @@ static int mkreg_ghost(char *path, GhostFileEntry *gfe, struct cr_img *img)
 	return ret;
 }
 
+static int mklnk_ghost(char *path, GhostFileEntry *gfe, struct cr_img *img)
+{
+	if (!gfe->symlnk_target) {
+		pr_err("Ghost symlink target is NULL for %s. Image from old CRIU?\n", path);
+		return -1;
+	}
+
+	if (symlink(gfe->symlnk_target, path) < 0) {
```

Comment:
What if the folder where the `symlnk_target` has been created no longer exists?

Reply:
> What if the folder of `path` no longer exists?

@rst0git Thank you for your point. If I understanding correctly, this case automatically handled because we call `mklnk_ghost` function only from `create_ghost` function. In which we already have a check for `ENOENT` error. It seems sufficient for me.

Reply:
In the case of `ENOENT`, CRIU will generate the following log message

```
Error (criu/files-reg.c:332): Could not create ghost symlink ... : No such file or directory
```
and execute:
```C
	/* Use grand parent, if parent directory does not exist */
	if (errno == ENOENT) {
		if (trim_last_parent(path) < 0) {
			pr_err("trim failed: @%s@\n", path);
			goto err;
		}
		goto again;
	}
```
As a result, it will try to create the symlink in the parent directory, potentially followed by successful restore. IMHO either the log message should not be an "error", or the restore should fail.

If the path in the ghost symlink is relative, would restoring it in a different directory make any difference?

Reply:
@rst0git thank you for the detailed explanation. I understood your point. I'm a little dumb %)
I've removed the error message and the behavior of `mklnk_ghost` should be the same as `mkreg_ghost` of course.

>If the path in the ghost symlink is relative, would restoring it in a different directory make any difference?

Great notice! But I think that in our case, it's really not important because we deal with ghost symlink and file descriptor opened with `O_PATH | O_NOFOLLOW` flags. In current kernels, all that we can do with that file descriptor is to `readlink()` / `fcntl()` / `fstat()` / `close()`.

Reply:
We discussed with Pavel (@Snorch) and after that discussion, I've decided to take a closer look at the potential problems with symlinks, created in the not original directory.

Idea of the test
1. Create symlink that points to file
2. open with `O_PATH | O_NOFOLLOW`
3. make symlink ghost (unlink it)
4. Open it using hacky procfs way with dereferencing: 
```
char lpath[1000];
sprintf(lpath, "/proc/self/fd/%d", fd);
ret = open(lpath, O_RDONLY);
if (ret == -1) {
	pr_perror("cat't open file");
	return 1;
}
```
5. And... we got a error:
`cat't open file (errno = 40 (Too many levels of symbolic links))`

All of that looks strange as for me. And this is a very weird and exotic way of using `O_PATH | O_NOFOLLOW` file descriptors, and also I've not found a way to crack the system. Perhaps,  this way exists but I don't know...